### ISSUE TITLE
Sending packets to other clients when handling packets

### ISSUE THREAD

**Issue by Ezrashaw** [Eastrall/LiteNetwork#49]
### Limitation

Currently, when handling packets in a `LiteServerUser` you can only send packets to that client. This doesn't allow for  'broadcasting' to other clients or directing messages at other clients.

https://github.com/Eastrall/LiteNetwork/blob/8deec63a86be9002b6d9c7dd657a5a4a0f2a51ea/src/LiteNetwork/Server/LiteServerUser.cs#L31

### Proposed Changes
Either 
1.  Extra methods could be added to `LiteServerUser` that allow sending packets to clients based on id or maybe predicates on the list of clients. See below for a potential broadcast methods.

```C#
/// <summary>
/// Broadcasts a packet to all clients (optionally excluding the current client)
/// </summary>
/// <param name="packetBuffer">The packet buffer to send</param>
/// <param name="excludeCurrentClient">Whether to exclude the current client in the broadcast</param>
public virtual void Broadcast(byte[] packetBuffer, bool excludeCurrentClient = true);

/// <summary>
/// Broadcasts a packet to all clients (optionally excluding the current client)
/// </summary>
/// <param name="packetBuffer">The packet stream to send</param>
/// <param name="excludeCurrentClient">Whether to exclude the current client in the broadcast</param>
public virtual void Broadcast(Stream packetStream, bool excludeCurrentClient = true);
```

2. Just adding a refence to the `LiteServer` in  `LiteServerUser` which personally I don't think is a very good solution as it doesn't allow for creating broadcasts to all clients _except the client currently handling the packet_ and handling more complex situations.

### Some extra ideas
- Groups of users which can each be sent messages individually (like rooms in [socket.io](https://socket.io/), documented [here](https://socket.io/docs/v4/rooms/))
- Maybe (optionally) allowing clients to move 'rooms' and adding handlers for `OnRoomChange` or something like that.
- Not strictly related to the issue, but adding `OnUserInitialize` in `LiteServer` to allow control over how `LiteServerUser`s are created (and just greater clarity around `LiteServerUser` constructors)

****
**Finally**
I really love the library, it's really nice and simple to use.
Really want to see proper packet serialization and just more great stuff lol.

Also, I realize there is `LiteConnection` and that is (probably) where these changes would be made (but I only just realized and can't be bothered to change everything).

**Comment by Eastrall** [comment]
Hi, thanks for opening a issue on `LiteNetwork`.
You can broadcast messages by injecting the `LiteServer<TUser>` inside the custom implementation of `LiteServerUser` and then call the `SendTo()` or `SendToAll()` methods: https://github.com/Eastrall/LiteNetwork/blob/main/src/LiteNetwork/Server/LiteServer.cs#L197

```csharp
public class CustomUser : LiteServerUser
{
    private readonly LiteServer<CustomUser> _server;

    public CustomUser(LiteServer<CustomUser> server)
    {
        _server = server;
    }

    public override Task HandleMessageAsync(byte[] packetBuffer)
    {
        // read packet
        _server.SendToAll(...);
    }
}
```

I'd like to react to your extra ideas too. The purpose of the library is to provide an API on top of the basic `Socket` API which can be tought to understand and come up with a fully TCP managed client or server. `LiteNetwork` solves this by providing a simple API for both client and server ; and also server-side user management.
The purpose of creating a room system shouldn't be inside the library but your own implementation. You could imagine a class named `Room` that contains a list of `CustomUser` (implementing the `LiteSeverUser`) and then have some methods to manage the room:
```csharp
public class Room
{
    private readonly List<CustomUser> _users = new();

    public void AddUser(CustomUser user)
    {
        // verify if user is already in room.
        _users.Add(user);
    }

    public void RemoveUser(CustomUser user)
    {
        // ...
    }

    public void BroadcastToAll(byte[] packetBuffer)
    {
         // broadcast to all players in room.
    }

    // ...
}
```

Coming back to the original topic, maybe we could think of a mechanism such as the `HttpContext` on ASP.NET Core, attached to the `LiteServerUser`. Thus, we this context, we might be able to access server configuration, maybe methods such as broadcasts and any other utility methods or properties.

I am glad you like the library! I put much effort on simplifying server and client creation in order to setup robust prototypes and real scenario client and servers! Thank you for contributing by proposing such changes, this always leads to constructive discussion!
As for the packet serialization, V1 had the `LitePacket` and `LitePacketStream` classes, but they have been removed in `v2.0.0-rc.[X]`. To write and read packets, you can use the traditional `BinaryWriter` and `BinaryReader`.
I have created a sample that shows how to create a custom packet reader and writer: https://github.com/Eastrall/LiteNetwork/tree/main/samples/custom-packet-reader-writer

**Comment by Ezrashaw** [comment]
Thanks for your response.

**Firstly**, I feel that DI is not really the best way to have such functionality, with the current implementation it feels clunky and people might not always want to use DI and I don't think DI should be forced upon people in that way.

 I do like your idea about ASP.NET Core's `HttpContext` and that could be a clean way to allow interacting with the `LiteServer<TUser>` from the `LiteServerUser`. It could implement methods like `Broadcast` and so on.

**Secondly**, I think that `Room`'s should be an important aspect of `LiteNetwork`.  `Room`'s are implemented in other networking libraries (like [socket.io](socket.io)) and are too basic to be left to the end user to implement. You said that `LiteNetwork` should provide server-side user management, `Room`'s are server side user management; they allow grouping and sorting users in a simple and clean way.

**Thirdly**, off topic but it would be really cool to see a system where you can define packets as classes (implementing an interface) and then send and receive those packets without using `byte[]` and `BinaryReader` which can be complex and seem like unnecessary boilerplate especially to new users. 

This is well implemented in a C# Minecraft Server called `Obsidian` ([here](https://github.com/ObsidianMC/Obsidian) and net code [here](https://github.com/ObsidianMC/Obsidian/tree/master/Obsidian/Net)), note the `MinecraftStream` which wraps the `NetworkStream` and provides methods to read/write basic types. They then use source gen to create `Serialize` and `Deserialize` (that use the `MinecraftStream`) on packet classes implementing an `IPacket` interface. I understand that source gen is probably out of scope for this project but perhaps a simpler system could be created using `Reflection` which would allow new users to easily create reasonably complex protocols with simple classes and code.

**Finally**, again love the library, thanks for the response.

**Comment by Eastrall** [comment]
I agree with you, Dependency Injection can be hard to understand for people that started developing not long ago. Nonetheless, I think it should still be part of the system, just like ASP.NET Core controllers for example.
I will start thinking about implementing a "Context" system for the `LiteServerUser` and maybe provide the following API:
```csharp
public class LiteServerUser : LiteConnection
{
    protected internal LiteServerContext Context { get; internal set; }
}
```
That could be used like this:
```csharp
public sealed class CustomUser : LiteServerUser
{
    public void SendChatMessage(string message)
    {
        using var packetStream = new MemoryStream();
        using var writer = new BinaryWriter(packetStream );

        writer.Write((byte)PacketType.ChatMessage); // PacketType is an enum that contains packet headers.
        writer.Write(message);

        Context.SendToAll(packetStream);
    }
}
```

Just like the room system, I believe the source generator process shouldn't be part of `LiteNetwork` **core library**, since `LiteNetwork` was designed to be a wrapper around TCP/IP sockets for creating server and clients easily.
This could in fact be **"extensions"**, to the core library, for example : `LiteNetwork.Extensions.Rooms` that uses `LiteServer` underneeth and provide an API to manage a room system. We could also imagine a  `LiteNetwork.Extensions.PacketSourceGenerator` that will implement a packet generator system, just like Obsidian. If you are interested, I used `LiteNetwork` on a side-project of recreation of a Minecraft server: https://github.com/Eastrall/Redstone. Each packets is in fact a class that takes the necessary parameters: https://github.com/Eastrall/Redstone/blob/main/src/Redstone.Protocol/Packets/Game/Client/SpawnPlayerPacket.cs
and I call it like this: https://github.com/Eastrall/Redstone/blob/78a32302c0829c60b4264326360a5968b46a0e7e/src/Redstone.Server/Entities/Player.cs#L125-L131

This could give you also some ideas about how to isolate a packet structure.

**Comment by Ezrashaw** [comment]
I think that all of these ideas are good.

- The `LiteServerContext` looks nice and I think we need to think about what methods `LiteServerContext` will implement and also about how the existing methods on both `LiteServer` and `LiteServerUser` will be changed. (outside of the scope of this issue)

- _Personally_, I feel that both the `Room` idea and the `PacketSourceGenerator` should be included in the core `LiteNetwork` library however the extension concept is a good compromise. 

- Your `Redstone`  project seems like a good example of generic packet serialization code that could be included in `LiteNetwork` (or an extension). `Redstone` seems quite similar `Obsidian` (in terms of its packet serialization code) except `Obsidian` uses Source Gen as opposed to the `MinecraftPacket` class directly calling methods on the `MinecraftStream` (I'm not which would be implemented in the `PacketSourceGenerator`)

**Finally**, I believe that simplicity and approachability to new users should be a key priority.

**Comment by Ezrashaw** [comment]
This issue is now closed with reference to #50.
